src/Request.php

```diff
@@ -352,18 +396,90 @@ protected function has_authentication_errors() {
 			$nonce = $_SERVER['HTTP_X_WP_NONCE']; // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash,WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
 		}
 
-		if ( null === $nonce ) {
-			// No nonce at all, so act as if it's an unauthenticated request.
+		/**
+		 * Treat "falsy" nonce values as "no nonce provided" rather than "invalid nonce".
+		 *
+		 * This handles common JavaScript serialization edge cases where null/undefined
+		 * values get converted to strings like "null" or "undefined", or where an empty
+		 * string is passed. These are clearly not real nonces and should be treated as
+		 * "no nonce" (downgrade to guest) rather than "invalid nonce" (error).
+		 */
+		$empty_nonce_values = [ '', 'null', 'undefined', 'false', '0' ];
+		if ( in_array( $nonce, $empty_nonce_values, true ) ) {
+			$nonce = null;
+		}
+
+		/**
+		 * Filter whether to require a nonce for cookie-based authentication.
+		 *
+		 * By default, WPGraphQL requires a nonce (X-WP-Nonce header or _wpnonce parameter)
+		 * for cookie-authenticated requests to prevent CSRF attacks. This matches the
+		 * behavior of the WordPress REST API.
+		 *
+		 * Note: WPGraphQL's CORS headers already mitigate most CSRF attack vectors by
+		 * blocking cross-origin requests with credentials. The nonce requirement provides
+		 * defense-in-depth protection, particularly against form-based CSRF attacks which
+		 * are not blocked by CORS.
+		 *
+		 * Developers may want to disable this requirement in local/development environments
+		 * to allow easier testing via browser URL bar or tools that don't send nonces.
+		 *
+		 * @since 2.4.0
+		 *
+		 * @param bool $require_nonce Whether to require a nonce for cookie auth. Default true.
+		 * @param \WPGraphQL\Request $request The current request instance.
+		 *
+		 * @return bool
+		 */
+		$require_nonce = apply_filters( 'graphql_cookie_auth_require_nonce', true, $this );
+
+		/**
+		 * No nonce provided - downgrade to unauthenticated (unless nonce requirement is disabled).
+		 *
+		 * This is not an error, we just treat the request as if it came from
+		 * an unauthenticated user (CSRF protection).
+		 */
+		if ( null === $nonce && $require_nonce ) {
 			wp_set_current_user( 0 );
```

Comment:
It still seems too late to call `wp_set_current_user`. Can this request-level authentication be done immediately _before_ `graphql_init`?

If _any_ WPGraphQL hooks run before this check, it feels like additional opportunity for mismatched expectations and potential leaks.

For example, `graphql_process_http_request` runs before this check, which plugins might hook into to enable sensitive debugging output.

Reply:
@chriszarate `graphql_init` is fired within the `after_setup_theme` callback, so adding something like this there would be overhead on _every_ WordPress request, even ones that are not making a GraphQL request. 

I believe we could move it to the Router and run it early there, as this is only protecting HTTP requests to the GraphQL endpoint. Having the code in Request meant that it was central to _all_ kinds of requests, including direct calls to the `graphql()` method which instantiates a `new \WPGraphQL\Request()` and fires it off. However, we're not protecting against direct calls to the graphql() method, just HTTP requests to the endpoint, so I think we can safely move this code early in the Router. 

I'll work on that refactor.